### PR TITLE
name kafka streams topology nodes appropriately

### PR DESCRIPTION
**Pull request by laxmanchekka**

### CODE REVIEW

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public List<String> getInputTopics(Map<String, Object> properties) {
       Config viewGenConfig = viewGenConfigs.get(viewGen);
       inputTopics.addAll(viewGenConfig.getStringList(INPUT_TOPICS_CONFIG_KEY));
     }
-    return inputTopics.stream().collect(Collectors.toList());
+    return new ArrayList<>(inputTopics);
```
Comment (aaron-steinfeld) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:69]: nit - why this change? List.copyOf (unless we're intentionally making this mutable)

That said, we should make this deterministic and avoid the hash set here.

(same comment below)

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java [hunk 2]
```diff
@@ -48,15 +43,14 @@ public void init(ProcessorContext context) {
...
-  public List<KeyValue<String, OUT>> transform(String key, IN value) {
-    List<KeyValue<String, OUT>> outputKVPairs = new ArrayList<>();
-    List<OUT> output = viewGenerator.process(value);
+  public void process(Record<String, IN> record) {
+    long nowMillis = Instant.now().toEpochMilli();
+    List<OUT> output = viewGenerator.process(record.value());
     if (output != null) {
       for (OUT out : output) {
-        outputKVPairs.add(KeyValue.pair(key, out));
+        context.forward(new Record<>(record.key(), out, nowMillis));
```
Comment (aaron-steinfeld) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java:51]: what's the millis input on the record represent?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java (hunk 1 above)
Comment (laxmanchekka) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:69]: > why this change?

Intellij suggestion

> avoid the hash set here

set needed as input topic across all viewgens is same and we need to add only one source per input topic.

Comment (aaron-steinfeld) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:69]: > set needed as input topic across all viewgens is same and we need to add only one source per input topic.

So we have a uniqueness requirement, not a set it sounds like. Using a hash set means that different runs with the same input will return their outputs in a different order. Even if that's not impactful today, it's a bad practice and makes writing tests annoying. Suggest any of
- use a list, than run distinct on it
- use an insertion-ordered set (e.g. guava's or linkedhashset I believe)
- use streams e.g.
```java
return getViewGenName(properties)
  .stream()
  .map(viewGenConfigs::get)
  .map(config -> config.getStringList(INPUT_TOPICS_CONFIG_KEY))
  .flatMap(Collection::stream)
  .distinct()
  .collect(Collectors.toUnmodifiableList());
```

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java (hunk 2 above)
Comment (laxmanchekka) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java:51]: record generation timestamp. will be useful later for the calculation of latencies.

image

Comment (aaron-steinfeld) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java:51]: So the output record - shouldn't its timestamp be the time it's forwarded (i.e. after we call the process function)? Could be I'm misunderstanding here though.

Comment (laxmanchekka) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGenerationProcessor.java:51]: > shouldn't its timestamp be the time it's forwarded

Yes. Will move the ts post process.

To be very precise and perfect, we can even generate a timestamp for output record.
But, I was just trying to save some unnecessary gc here by not creating a timestamp in a loop.

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java (hunk 1 above)
Comment (laxmanchekka) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:69]: Why not a TreeSet? 
Set for me ensures uniqueness and TreeSet ensures ordering (for tests thingy you mentioned).

Looks simpler and much readable than the streams snippet in the suggestion.

Comment (aaron-steinfeld) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:69]: > Looks simpler and much readable than the streams snippet in the suggestion.

matter of opinion! but yes, any ordered set will do for repeatability, doesn't need to be insertion ordered - I was just trying to match the perceived intent.

**Review by aaron-steinfeld — APPROVED** [review]

**Comment by laxmanchekka** [comment]
fixed the comments and vulnerabilities.

File: view-creator-framework/build.gradle.kts [hunk 3]
```diff
@@ -52,21 +52,21 @@ dependencies {
...
   implementation(platform("org.glassfish.jersey:jersey-bom:2.40"))
   implementation(platform("org.jetbrains.kotlin:kotlin-bom:1.6.21"))
 
-  compileOnly("org.projectlombok:lombok:1.18.24")
-  annotationProcessor("org.projectlombok:lombok:1.18.24")
-  implementation("org.slf4j:slf4j-api:1.7.36")
-  implementation("com.typesafe:config:1.4.1")
+  compileOnly("org.projectlombok:lombok:1.18.26")
+  annotationProcessor("org.projectlombok:lombok:1.18.26")
+  implementation("org.slf4j:slf4j-api:2.0.5")
```
Comment (satish-mittal) [inline view-creator-framework/build.gradle.kts:65]: After this change, hypertrace/hypertrace-ingestor transitively loads `slf4j-api:2.0.5` but is still on `log4j-slf4j-impl` instead of `slf4j2`. That led to failure of integration tests in QS.

Comment (satish-mittal) [reply view-creator-framework/build.gradle.kts:65]: Raised PR to fix it: https://github.com/hypertrace/hypertrace-ingester/pull/425
